src/i2c.rs

```diff
@@ -74,253 +74,219 @@ macro_rules! busy_wait {
-
-                        // put byte on the wire
-                        self.i2c.txdr.write(|w| { w.txdata().bits(*byte) });
-                    }
-
-                    // automatic STOP
-
-                    Ok(())
-                }
-            }
+impl <SCL, SDA> I2c<I2C2, (SCL, SDA)> {
+    pub fn i2c2<F>(
+        i2c: I2C2,
+        pins: (SCL, SDA),
+        freq: F,
+        clocks: Clocks,
+        apb1: &mut APB1R1,
+    ) -> Self where
+        F: Into<Hertz>,
+        SCL: SclPin<I2C2>,
+        SDA: SdaPin<I2C2>,
+    {
+        apb1.enr().modify(|_, w| w.i2c2en().set_bit());
+        apb1.rstr().modify(|_, w| w.i2c2rst().set_bit());
+        apb1.rstr().modify(|_, w| w.i2c2rst().clear_bit());
+        Self::new(i2c, pins, freq, clocks)
+    }
+}
 
-            impl<PINS> Read for I2c<$I2CX, PINS> {
-                type Error = Error;
-
-                fn read(&mut self,
-                    addr: u8,
-                    buffer: &mut [u8],) -> Result<(), Error> {
-                    self.i2c.cr2.write(|w| {
-                        w.sadd()
-                            .bits((addr as u16) << 1)
-                            .rd_wrn()
-                            .set_bit()
-                            .nbytes()
-                            .bits(buffer.len() as u8)
-                            .start()
-                            .set_bit()
-                            .autoend()
-                            .set_bit()
-                    });
-
-                    for byte in buffer {
-                        // Wait until we have received something
-                        busy_wait!(self.i2c, rxne);
-
-                        *byte = self.i2c.rxdr.read().rxdata().bits();
-                    }
-
-                    Ok(())
-                }
-            }
+impl<SCL, SDA, I2C> I2c<I2C, (SCL, SDA)> where I2C: Deref<Target = i2c1::RegisterBlock> {
+    /// Configures the I2C peripheral to work in master mode
+    fn new<F>(
+        i2c: I2C,
+        pins: (SCL, SDA),
+        freq: F,
+        clocks: Clocks,
+    ) -> Self where
+        F: Into<Hertz>,
+        SCL: SclPin<I2C>,
+        SDA: SdaPin<I2C>,
+    {
+        let freq = freq.into().0;
+        assert!(freq <= 1_000_000);
+
+        // Make sure the I2C unit is disabled so we can configure it
+        i2c.cr1.modify(|_, w| w.pe().clear_bit());
+
+        let _ = clocks;
+        const STM32CUBEMX_I2CTIMINGTRAITEMENT: u32 = 0x10E09CE6u32;
+        const TIMING_CLEAR_MASK: u32 = 0xF0FFFFFFu32;
+        // Configure
+        i2c.timingr.write(|w| {
+            unsafe { w.bits(STM32CUBEMX_I2CTIMINGTRAITEMENT & TIMING_CLEAR_MASK) }
+        });
+        // Check if the value matches.
+        assert_eq!(STM32CUBEMX_I2CTIMINGTRAITEMENT & TIMING_CLEAR_MASK, i2c.timingr.read().bits());
+        // Enable the peripheral
+        i2c.cr1.write(|w| w.pe().set_bit());
+
+        // Set interface address in 7 bit mode
+        i2c.oar1.write(|w| w.oa1en().set_bit());
```

Comment:
This will not change anything, as oa is short for 'own address', and thus is only relevant when using the peripheral in slave mode.

Reply:
Yeah I confirmed no effect. I drop this. Through the commits of this branch I'm trying to replecate what STMCubeMx does.